lib/adapters/class-vip-enterprise-search.php

```diff
@@ -197,6 +255,168 @@ public function filter__vip_search_post_taxonomies_allow_list( $taxonomies ) {
 		return $this->get_restricted_taxonomies() ?: $taxonomies;
 	}
 
+	/**
+	 * A callback for the wp_rest_search_handlers filter. Includes the
+	 * search-suggestions REST API handler in the list of allowed handlers if
+	 * search suggestions are made available over REST.
+	 *
+	 * @param \WP_REST_Search_Handler[] $search_handlers List of search handlers to use in the controller.
+	 * @return \WP_REST_Search_Handler[] Updated list of handlers.
+	 */
+	public function filter__wp_rest_search_handlers( $search_handlers ) {
+		if ( $this->is_show_search_suggestions_in_rest_enabled() ) {
+			$search_handlers[] = new Post_Suggestion_Search_Handler( $this );
+		}
+
+		return $search_handlers;
+	}
+
+	/**
+	 * Suggest posts that match the given search term.
+	 *
+	 * @param string $search Search string.
+	 * @param array  $args   {
+	 *     Optional. An array of arguments.
+	 *
+	 *     @type string[] $subtypes Limit suggestions to this subset of all post
+	 *                              types that support search suggestions.
+	 *     @type int      $page     Page of results.
+	 *     @type int      $per_page Results per page. Default 10.
+	 *     @type int[]    $include  Search within these post IDs.
+	 *     @type int[]    $exclude  Exclude these post IDs from results.
+	 *     @type string[] $status   Post statuses to search. Default 'publish'.
+	 * }
+	 * @return int[] Post IDs in this page of results and total number of results.
+	 */
+	public function query_post_suggestions( string $search, array $args = [] ): array {
+		$out = [ [], 0 ];
+
+		if ( ! $search && ! $this->get_allow_empty_search() ) {
+			return $out;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			[
+				'subtypes' => [],
+				'page'     => 1,
+				'per_page' => 10,
+				'exclude'  => [], // phpcs:ignore WordPressVIPMinimum.Performance.WPQueryParams.PostNotIn_exclude
+				'include'  => [],
+				'status'   => [ 'publish' ],
+			],
+		);
+
+		$subtypes = (array) $args['subtypes'];
+		$page     = (int) $args['page'];
+		$per_page = (int) $args['per_page'];
+		$exclude  = (array) $args['exclude'];
+		$include  = (array) $args['include'];
+
+		if ( $search ) {
+			$query_from     = ( $page - 1 ) * $per_page;
+			$query_per_page = $per_page;
+			$query_must     = [
+				[
+					'multi_match' => [
+						'query'    => $search,
+						'type'     => 'bool_prefix',
+						'fields'   => [
+							'search_suggest',
+							'search_suggest._2gram',
+							'search_suggest._3gram',
+						],
+
+						/*
+						 * The purpose of using 'and' here is to assume that more terms in the search indicate
+						 * someone searching for something more specific, not someone trying to get more results.
+						 */
+						'operator' => 'and',
```

Comment:
Suggestion: it'd be good if that's filterable.

Reply:
Not opposed to that, but I think I'd be in favor of waiting for someone to present a use case (unless you have one already!) so that we could be more confident that the filter provided the information they would need to make a decision.

Reply:
Agree! A lot of my feedback in this pr might have been overkill honestly (without a proper use case).